upload-api/tables/provisions.js

```diff
@@ -0,0 +1,188 @@
+}
+
+class ConflictError extends Failure {
+  /**
+   * @param {object} input
+   * @param {string} input.message
+   */
+  constructor({ message }) {
+    super(message)
+    this.name = 'ConflictError'
+  }
+}
+
+/**
+ * @param {DynamoDBClient} dynamoDb
+ * @param {string} tableName
+ * @param {import('@ucanto/interface').DID<'web'>[]} services
+ * @returns {import('@web3-storage/upload-api').ProvisionsStorage}
+ */
+export function useProvisionsTable (dynamoDb, tableName, services) {
+  return {
+    services,
+    hasStorageProvider: async (consumer) => {
+      const cmd = new QueryCommand({
+        TableName: tableName,
+        KeyConditions: {
+          consumer: {
+            ComparisonOperator: 'EQ',
+            AttributeValueList: [{ S: consumer }]
+          }
+        },
+        AttributesToGet: ['cid']
+      })
+      const response = await dynamoDb.send(cmd)
+      const itemCount = response.Items?.length || 0
+      return { ok: itemCount > 0 }
+    },
+    /**
+     * ensure item is stored
+     *
+     * @param item - provision to store
+     */
+    put: async (item) => {
+      const row = {
+        cid: item.cause.cid.toString(),
+        consumer: item.consumer,
+        provider: item.provider,
+        customer: item.customer,
+      }
+      try {
+        await dynamoDb.send(new PutItemCommand({
+          TableName: tableName,
+          Item: marshall(row),
+          ConditionExpression: `attribute_not_exists(consumer) OR ((cid = :cid) AND (consumer = :consumer) AND (provider = :provider) AND (customer = :customer))`,
+          ExpressionAttributeValues: {
+            ':cid': { 'S': row.cid },
+            ':consumer': { 'S': row.consumer },
+            ':provider': { 'S': row.provider },
+            ':customer': { 'S': row.customer }
+          }
+        }))
+      } catch (error) {
+        if (error instanceof Error && error.message === 'The conditional request failed') {
+          return {
+            error: new ConflictError({
+              message: `Space ${row.consumer} cannot be provisioned with ${row.provider}: it already has a provider`
+            })
+          }
+        } else {
+          throw error
+        }
+      }
+      return { ok: {} }
+    },
+
+    /**
+     * get number of stored items
+     */
+    count: async () => {
+      const result = await dynamoDb.send(new DescribeTableCommand({
+        TableName: tableName
+      }))
+
+      return BigInt(result.Table?.ItemCount ?? -1)
+    }
+  }
+}
+
+/**
+ * @param {DynamoDBClient} dynamoDb
+ * @param {string} subscriptionsTableName
+ * @param {string} consumersTableName
+ * @param {import('@ucanto/interface').DID<'web'>[]} services
+ * @returns {import('@web3-storage/upload-api').ProvisionsStorage}
+ */
+export function useProvisionsStorage (dynamoDb, subscriptionsTableName, consumersTableName, services) {
+  return {
+    services,
+    hasStorageProvider: async (consumer) => {
+      const cmd = new QueryCommand({
```

Comment:
Same as below. Perhaps just get?

Reply:
removed this in favor of subscription/consumer tables